src/disco/shred/fd_fec_resolver.c

```diff
@@ -0,0 +1,527 @@
+      * no key is in both the done tcache and the current map
+      * each set pointer provided to the new function is in exactly one
+          of curr_map, freelist, or complete_list
+      * bmtree_free_list has exactly partial_depth fewer elements than
+          freelist
+   */
+
+  int dup = 0;
+  (void)dup;
+  /* Note: we identify FEC sets by the first 64 bits of their signature.
+     Given how slow Ed25519 is and how short of a time these are
+     relevant for, this seems safe, but we should research the issue
+     further. */
+
+  /* Are we already done with this FEC set? */
+  /* TODO: xor this with something semi-secret to prevent someone from
+     spending a lot of compute to make a shred with a 0 signature which
+     would be interpreted as an invalid key. */
+  ulong signature = fd_ulong_load_8( shred->signature );
+  int found;
+  ulong map_idx;
+
+  (void)map_idx;
+  FD_TCACHE_QUERY( found, map_idx, done_map, done_map_cnt, signature );
+  if( found )  return FD_FEC_RESOLVER_SHRED_IGNORED; /* With no packet loss, we expect found==1 about 50% of the time */
+
+  set_ctx_t * ctx = ctx_map_query( curr_map, signature, NULL );
+
+  fd_bmtree_node_t _leaf[1];
+  uchar variant    = shred->variant;
+  uchar shred_type = fd_shred_type( variant );
+
+  ulong tree_depth           = fd_shred_merkle_cnt( variant );
+  ulong reedsol_protected_sz = 1115UL - 20UL*tree_depth + 0x58UL - 0x40UL;
+  ulong merkle_protected_sz  = reedsol_protected_sz + fd_ulong_if( shred_type==FD_SHRED_TYPE_MERKLE_DATA, 0UL, 0x59UL - 0x40UL );
+
+  fd_bmtree_hash_leaf( _leaf, (uchar const *)shred + sizeof(fd_ed25519_sig_t), merkle_protected_sz, FD_BMTREE_LONG_PREFIX_SZ );
+
+  /* code_idx: if this is a Merkle coding shred, the index of it in the
+     Merkle array, otherwise 0. */
+  ulong code_idx   = fd_ulong_if( shred_type==FD_SHRED_TYPE_MERKLE_DATA, 0UL, shred->code.idx );
+  ulong shred_idx  = fd_ulong_if( shred_type==FD_SHRED_TYPE_MERKLE_DATA,
+                                  shred->idx - shred->fec_set_idx,
+                                  shred->code.data_cnt + code_idx );
+  /* TODO: bounds check shred_pos. make sure proof depth has the right
+       value */
+
+  if( FD_UNLIKELY( !ctx ) ) {
+    /* This is the first shred in the FEC set */
+    if( FD_UNLIKELY( freelist_cnt( free_list )<=partial_depth ) ) {
+      /* Packet loss is really high and we have a lot of in-progress FEC
+         sets that we haven't been able to finish.  Take the resources
+         bmtree from the oldest, and send the oldest FEC set to the back
+         of the free list. */
+      ulong oldest_key = resolver->curr_tail_tag;
+      set_ctx_t * victim_ctx = ctx_map_query( curr_map, oldest_key, NULL );
+      if( FD_UNLIKELY( !victim_ctx ) ) FD_LOG_ERR(( "data structures not in sync. Couldn't find %lx in the map", oldest_key ));
+      resolver->curr_tail_tag = victim_ctx->prev_tag;
+      ctx_map_query( curr_map, victim_ctx->prev_tag, NULL )->next_tag = 0UL;
```

Comment:
Also check for corruption here if ctx_map_query is NULL ?

Reply:
This is also gone now